File: apps/mobile/lib/nutrition.ts
import type { MacroLevel, NutritionCycleType, NutritionPhase } from "@/lib/api";

export const NUTRITION_CYCLE_TYPE_OPTIONS: ReadonlyArray<{
  value: NutritionCycleType;
  label: string;
}> = [
  { value: "macro_profile", label: "Macro Profile" },
  { value: "named_diet", label: "Named Diet" },
  { value: "elimination", label: "Elimination" },
  { value: "custom", label: "Custom" },
];

export const MACRO_LEVEL_OPTIONS: ReadonlyArray<{ value: MacroLevel; label: string }> = [
  { value: "low", label: "Low" },
  { value: "medium", label: "Medium" },
  { value: "high", label: "High" },
];

export function getNutritionCycleTypeLabel(value: NutritionCycleType): string {
  return NUTRITION_CYCLE_TYPE_OPTIONS.find((option) => option.value === value)?.label ?? value;
}

export function formatMacroProfile(
  macroProfile: NutritionPhase["macro_profile"] | null | undefined
): string | null {
  if (!macroProfile) {
    return null;
  }

  return `Carbs ${capitalize(macroProfile.carbs)} · Protein ${capitalize(
    macroProfile.protein
  )} · Fat ${capitalize(macroProfile.fat)}`;
}

export function formatNutritionPhaseSummary(phase: NutritionPhase): string[] {
  const summary: string[] = [];
  const macros = formatMacroProfile(phase.macro_profile);
  if (macros) {
    summary.push(macros);
  }
  if (phase.pattern?.trim()) {
    summary.push(phase.pattern.trim());
  }
  if (phase.restrictions.length > 0) {
    summary.push(phase.restrictions.join(", "));
  }
  return summary;
}

function capitalize(value: string): string {
  return value.charAt(0).toUpperCase() + value.slice(1);
}

File: apps/api/app/schemas/nutrition.py
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

NutritionCycleType = Literal["macro_profile", "named_diet", "elimination", "custom"]
MacroLevel = Literal["low", "medium", "high"]


class NutritionMacroProfile(BaseModel):
    carbs: MacroLevel
    protein: MacroLevel
    fat: MacroLevel


class NutritionPhasePayload(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    duration_days: int = Field(gt=0, le=365)
    macro_profile: NutritionMacroProfile | None = None
    pattern: str | None = Field(default=None, max_length=120)
    restrictions: list[str] = Field(default_factory=list)
    notes: str | None = None

    @model_validator(mode="after")
    def validate_phase_shape(self):
        has_macro_profile = self.macro_profile is not None
        has_pattern = bool(self.pattern and self.pattern.strip())
        has_restrictions = any(item.strip() for item in self.restrictions)
        has_notes = bool(self.notes and self.notes.strip())

        if not (has_macro_profile or has_pattern or has_restrictions or has_notes):
            raise ValueError("Each nutrition phase needs macro targets, a named pattern, restrictions, or notes")
        return self


class NutritionCycleCreate(BaseModel):
    cycle_type: NutritionCycleType
    name: str = Field(min_length=1, max_length=255)
    phases: list[NutritionPhasePayload] = Field(min_length=1, max_length=12)
    phase_started_at: date


class NutritionCycleUpdate(BaseModel):
    cycle_type: NutritionCycleType | None = None
    name: str | None = Field(default=None, min_length=1, max_length=255)
    phases: list[NutritionPhasePayload] | None = Field(default=None, min_length=1, max_length=12)
    phase_started_at: date | None = None
    is_active: bool | None = None


class NutritionPhaseResponse(NutritionPhasePayload):
    pass


class ActiveNutritionPhaseResponse(NutritionPhaseResponse):
    plan_name: str
    cycle_type: NutritionCycleType
    current_phase_idx: int
    total_phases: int
    next_transition: date
    days_until_transition: int


class NutritionCycleResponse(BaseModel):
    id: uuid.UUID
    cycle_type: NutritionCycleType
    name: str
    phases: list[NutritionPhaseResponse]
    current_phase_idx: int
    phase_started_at: date
    next_transition: date
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}
